DataManagementSystem/Agent/FTS3Agent.py

```diff
@@ -0,0 +1,420 @@
+
+    contextes = self._globalContextCache.setdefault( threadID, DictCache() )
+    
+    idTuple =(username, group, ftsServer) 
+    log.debug( "Getting context for %s" % ( idTuple, ) )
+
+    if not contextes.exists( idTuple, 2700 ):
+      res = getDNForUsername(username)
+      if not res['OK']:
+        return res
+      # We take the first DN returned
+      userDN = res['Value'][0]
+      
+      log.debug( "UserDN %s" % userDN )
+
+      # We dump the proxy to a file.
+      # It has to have a lifetime of at least 2 hours
+      # and we cache it for 1.5 hours
+      res = gProxyManager.downloadVOMSProxyToFile(userDN,
+                                                  group,
+                                                  requiredTimeLeft = 7200,
+                                                  cacheTime = 5400 )
+      if not res['OK']:
+        return res
+      
+      proxyFile = res['Value']
+      log.debug( "Proxy file %s" % proxyFile )
+      
+      # We generate the context
+      res = FTS3Job.generateContext( ftsServer, proxyFile )
+      if not res['OK']:
+        return res
+      context = res['Value']
+
+      # we add it to the cache for this thread for 1h
+      contextes.add( idTuple, 3600, context )
+
+    return S_OK( contextes.get( idTuple ) )
+
+
+  
+  
+  def _monitorJob( self, ftsJob ):
+    """
+        * query the FTS servers
+        * update the FTSFile status
+        * update the FTSJob status
+    """
+    # General try catch to avoid that the tread dies
+    try:
+      threadID = current_process().name
+      log = gLogger.getSubLogger( "_monitorJob/%s" % ftsJob.jobID, child = True )
+  
+      res = self.getFTS3Context( ftsJob.username, ftsJob.userGroup, ftsJob.ftsServer, threadID = threadID )
+  
+      if not res['OK']:
+        log.error( "Error getting context", res )
+        return ftsJob, res
+  
+      context = res['Value']
+  
+  
+      res = ftsJob.monitor( context = context )
+  
+      if not res['OK']:
+        log.error( "Error monitoring job", res )
+        return ftsJob, res
+  
+      # { fileID : { Status, Error } }
+      filesStatus = res['Value']
+  
+      res = self.fts3db.updateFileStatus( filesStatus )
+  
+      if not res['OK']:
+        log.error( "Error updating file fts status", "%s, %s" % ( ftsJob.ftsGUID, res ) )
+        return ftsJob, res
+  
+
+      upDict = {ftsJob.jobID :  {'status' : ftsJob.status,
+                                                              'error' : ftsJob.error,
+                                                              'completeness' : ftsJob.completeness,
+                                                              'operationID' : ftsJob.operationID,
+                                                              }
+                                          }
+      res = self.fts3db.updateJobStatus( upDict )
+
+      return ftsJob, res
+
+    except Exception as e:
+      return ftsJob, S_ERROR( "Exception %s" % repr( e ) )
```

Comment:
DErrno?

Reply:
Delicate for such a general exception.
I added DErrno where possible now